File: ideaproject/javaserializer/src/com/houseelectrics/serializer/StringUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace com.houseelectrics.serializer
{
    
    public static class StringUtils
    {
        public delegate bool Match(char c, string str, int pos);

        public static int? firstCharPosition(this string str, Match match, int startPos = 0, int? endPos = 0)
        {
            if (endPos == null) endPos = str.Length;
            for (int pos = startPos; pos < endPos + 1; pos++)
            {
                if (match(str[pos], str, pos)) return pos;
            }
            return null;
        }
        
        public static int? firstCharPosition(this string str, Predicate<char> match, int startPos = 0, int? endPos = 0)
        {
            if (endPos == null) endPos = str.Length;
            for (int pos = startPos; pos < endPos+1; pos++)
            {
                if (match(str[pos])) return pos;
            }
            return null;
        }

        public static int? lastCharPosition(this string str, Predicate<char> match, int? startPos = null, int endPos = 0)
        {
            if (startPos == null) startPos = str.Length;
            for (int pos = startPos.Value-1; pos >= endPos; pos--)
            {
                if (match(str[pos])) return pos;
            }
            return null;
        }
    
    }

     





}

File: ideaproject/javaserializer/src/com/houseelectrics/serializer/StringUtils.java
package com.houseelectrics.serializer;

/**
 * Created by roberttodd on 29/11/2014.
 */
public class StringUtils
{

    public interface Matcher
    {
        boolean match(char c, String str, int pos);
    }

    public static Integer firstCharPosition(String str, Matcher matcher, Integer startPos, Integer endPos)
{
    if (endPos == null) endPos = str.length();
    for (int pos = startPos; pos < endPos + 1; pos++)
    {
        if (matcher.match(str.charAt(pos), str, pos)) return pos;
    }
    return null;
}

    public interface CharPredicate
    {
        public boolean match(char c);
    }

    public static Integer firstCharPosition(String str, CharPredicate match)
    {
        return firstCharPosition(str, match, 0, 0);
    }

    public static Integer firstCharPosition(String str, CharPredicate match, int startPos, Integer endPos)
{
    if (endPos == null) endPos = str.length();
    for (int pos = startPos; pos < endPos+1; pos++)
    {
        if (match.match (str.charAt(pos))) return pos;
    }
    return null;
}

    public static Integer lastCharPosition(String str, CharPredicate match)
    {
        return lastCharPosition(str, match, null, 0);
    }


    public static Integer lastCharPosition(String str, CharPredicate match, Integer startPos, int endPos)
{
    if (startPos == null) startPos = str.length();
    for (int pos = startPos.intValue()-1; pos >= endPos; pos--)
    {
        if (match.match(str.charAt(pos))) return pos;
    }
    return null;
}


}
